File: 23. Function pointer demo.cpp
// Topic: Function pointer demo and array of function pointers (calculator)
#include <iostream>
using namespace std;

int add(int a, int b)
{
    return a + b;
}
int sub(int a, int b)
{
    return a - b;
}
int mul(int a, int b)
{
    return a * b;
}
int divide_(int a, int b)
{
    return b ? a / b : 0;
}

int operate(int (*op)(int, int), int a, int b)
{
    return op(a, b);
}

int main()
{
    int (*ops[4])(int, int) = {add, sub, mul, divide_};
    cout << "Add via ops[0]: " << ops[0](4, 2) << '\n';
    cout << "Operate(callback): " << operate(mul, 3, 5) << '\n';
    return 0;
}

// Time Complexity: O(1) for calls; function pointer overhead negligible.
// Space Complexity: O(1).

File: 23. Function pointer demo.py
"""
Function pointer demo in Python: use functions as first-class objects
"""
def add(a,b): return a+b
def sub(a,b): return a-b
def mul(a,b): return a*b
def div(a,b): return a//b if b else 0

def operate(op, a, b):
    return op(a,b)

def main():
    ops = [add, sub, mul, div]
    print('add via ops[0]:', ops[0](4,2))
    print('operate(callback):', operate(mul, 3, 5))

if __name__ == '__main__':
    main()

# Time Complexity: O(1) for individual calls.
# Space Complexity: O(1).
